flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PrioritizedOperatorSubtaskState.java

```diff
@@ -0,0 +1,265 @@
+ * "ground truth" about which state should be represented. Alternatives may be complete or partial substitutes for
+ * the "ground truth" with a higher priority (if they had a lower alternative, they would not really be alternatives).
+ * Substitution is determined on a per-sub-state basis.
+ */
+public class PrioritizedOperatorSubtaskState {
+
+	/** Singleton instance for an empty, non-restored operator state. */
+	private static final PrioritizedOperatorSubtaskState EMPTY_NON_RESTORED_INSTANCE =
+		new PrioritizedOperatorSubtaskState(new OperatorSubtaskState(), Collections.emptyList(), false);
+
+	/** List of prioritized snapshot alternatives for managed operator state. */
+	private final List<StateObjectCollection<OperatorStateHandle>> prioritizedManagedOperatorState;
+
+	/** List of prioritized snapshot alternatives for raw operator state. */
+	private final List<StateObjectCollection<OperatorStateHandle>> prioritizedRawOperatorState;
+
+	/** List of prioritized snapshot alternatives for managed keyed state. */
+	private final List<StateObjectCollection<KeyedStateHandle>> prioritizedManagedKeyedState;
+
+	/** List of prioritized snapshot alternatives for raw keyed state. */
+	private final List<StateObjectCollection<KeyedStateHandle>> prioritizedRawKeyedState;
+
+	/** Signal flag if this represents state for a restored operator. */
+	private final boolean restored;
+
+	public PrioritizedOperatorSubtaskState(
+		@Nonnull OperatorSubtaskState jobManagerState,
+		@Nonnull List<OperatorSubtaskState> alternativesByPriority) {
+		this(jobManagerState, alternativesByPriority, true);
+	}
+
+	public PrioritizedOperatorSubtaskState(
+		@Nonnull OperatorSubtaskState jobManagerState,
+		@Nonnull List<OperatorSubtaskState> alternativesByPriority,
+		boolean restored) {
+
+		Preconditions.checkNotNull(jobManagerState, "Job manager state is null.");
+		int size = Preconditions.checkNotNull(alternativesByPriority, "Alternative states are null.").size();
+
+		this.restored = restored;
+
+		List<StateObjectCollection<OperatorStateHandle>> managedOperatorAlternatives = new ArrayList<>(size);
+		List<StateObjectCollection<KeyedStateHandle>> managedKeyedAlternatives = new ArrayList<>(size);
+		List<StateObjectCollection<OperatorStateHandle>> rawOperatorAlternatives = new ArrayList<>(size);
+		List<StateObjectCollection<KeyedStateHandle>> rawKeyedAlternatives = new ArrayList<>(size);
+
+		for (OperatorSubtaskState subtaskState : alternativesByPriority) {
+
+			if (subtaskState != null) {
+				managedKeyedAlternatives.add(subtaskState.getManagedKeyedState());
+				rawKeyedAlternatives.add(subtaskState.getRawKeyedState());
+				managedOperatorAlternatives.add(subtaskState.getManagedOperatorState());
+				rawOperatorAlternatives.add(subtaskState.getRawOperatorState());
+			}
+		}
+
+		// Key-groups should match.
+		BiFunction<KeyedStateHandle, KeyedStateHandle, Boolean> keyedStateApprover =
+			(ref, alt) -> ref.getKeyGroupRange().equals(alt.getKeyGroupRange());
+
```

Comment:
Why `ref.getKeyGroupRange()` and `alt. ref.getKeyGroupRange ()` need to be strictly equal? It seems that `alt. ref.getKeyGroupRange ().begin <= ref. ref.getKeyGroupRange ().begin && alt. ref.getKeyGroupRange ().end >= ref ref.getKeyGroupRange().end` is also acceptable.

Reply:
In theory, we could be less strict, yes. But in practice it doesn't matter because rescaling is not supported right now. See my other comment.